protocol/messenger_status_updates.go

```diff
@@ -0,0 +1,247 @@
+package protocol
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"go.uber.org/zap"
+
+	"github.com/status-im/status-go/protocol/common"
+	"github.com/status-im/status-go/protocol/communities"
+	"github.com/status-im/status-go/protocol/protobuf"
+	"github.com/status-im/status-go/protocol/transport"
+)
+
+func (m *Messenger) GetCurrentUserStatus() (*UserStatus, error) {
+
+	status := &UserStatus{
+		StatusType: int(protobuf.StatusUpdate_ONLINE),
+		Clock:      0,
+		CustomText: "",
+	}
+
+	err := m.settings.GetCurrentStatus(status)
+	if err != nil {
+		m.logger.Debug("Error obtaining latest status", zap.Error(err))
+		return nil, err
+	}
+
+	return status, nil
+}
+
+func (m *Messenger) sendUserStatus(status UserStatus) error {
+	shouldBroadcastUserStatus, err := m.settings.ShouldBroadcastUserStatus()
+	if err != nil {
+		return err
+	}
+
+	if !shouldBroadcastUserStatus {
+		m.logger.Debug("user status should not be broadcasted")
+		return nil
+	}
+
+	status.Clock = uint64(time.Now().Unix())
+
+	err = m.settings.SaveSetting("current-user-status", status)
+	if err != nil {
+		return err
+	}
+
+	statusUpdate := &protobuf.StatusUpdate{
+		Clock:      status.Clock,
+		StatusType: protobuf.StatusUpdate_StatusType(status.StatusType),
+		CustomText: status.CustomText,
+	}
+
+	encodedMessage, err := proto.Marshal(statusUpdate)
+	if err != nil {
+		return err
+	}
+
+	contactCodeTopic := transport.ContactCodeTopic(&m.identity.PublicKey)
+
+	rawMessage := common.RawMessage{
+		LocalChatID:         contactCodeTopic,
+		Payload:             encodedMessage,
+		MessageType:         protobuf.ApplicationMetadataMessage_STATUS_UPDATE,
+		ResendAutomatically: true,
+	}
+
+	_, err = m.sender.SendPublic(context.Background(), contactCodeTopic, rawMessage)
+	if err != nil {
+		return err
+	}
+
+	joinedCommunities, err := m.communitiesManager.Joined()
+	if err != nil {
+		return err
+	}
+	for _, community := range joinedCommunities {
+		rawMessage.LocalChatID = community.StatusUpdatesChannelID()
+		_, err = m.sender.SendPublic(context.Background(), rawMessage.LocalChatID, rawMessage)
```

Comment:
you create many background contexts here in this function.
since they all belong to the same request, it shall be the same context instance passed as a function parameter.
this, in particular, ensures better unit testing.